File: addons/guide-cpp/src/modifiers/guide_modifier_window_relative.cpp
#include "guide_modifier_window_relative.h"
#include <godot_cpp/classes/engine.hpp>
#include <godot_cpp/classes/scene_tree.hpp>
#include <godot_cpp/classes/viewport.hpp>
#include <godot_cpp/classes/window.hpp>

using namespace godot;

Vector3 GUIDEModifierWindowRelative::_modify_input(Vector3 input, double delta, GUIDEAction::GUIDEActionValueType value_type) const {
    if (!input.is_finite()) return Vector3(Math_INF, Math_INF, Math_INF);

    SceneTree *tree = Object::cast_to<SceneTree>(Engine::get_singleton()->get_main_loop());
    if (!tree) return Vector3(Math_INF, Math_INF, Math_INF);

    Window *window = tree->get_root();
    if (!window) return Vector3(Math_INF, Math_INF, Math_INF);

    Vector2 size = window->get_screen_transform().affine_inverse().xform(window->get_size());
    return Vector3(input.x / size.x, input.y / size.y, input.z);
}

File: addons/GuideCS/Modifiers/GUIDEModifierWindowRelative.cs
using System;
using Godot;

namespace Guide;

[Tool]
[GlobalClass]
public partial class GUIDEModifierWindowRelative : GUIDEModifier {
	public override bool IsSameAs(GUIDEModifier other) => other is GUIDEModifierWindowRelative;

	public override Vector3 _ModifyInput(Vector3 input, float delta, GUIDEAction.GUIDEActionValueType valueType) {
		if (!input.IsFinite()) return Vector3.Inf;

		if (Engine.GetMainLoop() is not SceneTree tree) return Vector3.Inf;
		var window = tree.Root;
		var windowSize = window.GetScreenTransform().AffineInverse() * window.Size;
		return new Vector3(input.X / windowSize.X, input.Y / windowSize.Y, input.Z);
	}

	public override string _EditorName() => "Window relative";

	public override string _EditorDescription() => "Converts the value of the input into window-relative units between 0 and 1.\nE.g. if a mouse cursor moves half a screen to the right and down, then \nthis modifier will return (0.5, 0.5).";
}
